File: scanner/app/src/main/java/com/claycode/scanner/topology_decoder/BitTreeConverter.kt
/* MIT License with Commons Clause
Copyright (c) 2025 Claycode
See LICENSE file in the root of this project for license details.
Commercial use is prohibited without a separate license. */

package com.claycode.scanner.topology_decoder

import android.os.Build
import androidx.annotation.RequiresApi
import com.claycode.scanner.data_structures.BitString
import com.claycode.scanner.data_structures.Tree
import java.math.BigInteger

/**
 * This class provides conversions between bits and trees.
 * Bits are represented as strings of zeros and ones.
 */
class BitTreeConverter {
    companion object {
        fun bitsToTree(bits: BitString): Tree {
            @RequiresApi(Build.VERSION_CODES.TIRAMISU)
            fun populateTreeOfNumber(root: Tree, n: BigInteger) {
                val dec = numberToSquareDecomposition(n)
                root.children = dec.map { Tree() }.toMutableList()
                root.children.forEachIndexed { i, c ->
                    populateTreeOfNumber(c, dec[i])
                }
            }

            val root = Tree()
            val n = bitStringToNumber(bits)
            populateTreeOfNumber(root, n)

            return root
        }

        fun treeToBits(root: Tree): BitString {
            fun treeToNumber(root: Tree): BigInteger {
                if (root.children.isEmpty()) {
                    return BigInteger.ONE
                } else {
                    val dec = root.children.map { treeToNumber(it) }
                    return squareDecompositionToNumber(dec)
                }
            }

            return numberToBitString(treeToNumber(root))
        }

        private fun bitStringToNumber(bits: BitString): BigInteger {
            // Prepend 1 to make sure it is an integer
            return BigInteger("1${bits.toString()}", 2)
        }

        private fun numberToBitString(number: BigInteger): BitString {
            // Remove initial 1
            return BitString(number.toString(2).substring(1))
        }

        @RequiresApi(Build.VERSION_CODES.TIRAMISU)
        private fun largestSquare(n: BigInteger): BigInteger {
            if (n == BigInteger.ONE) return BigInteger.ONE

            var a = BigInteger.ZERO
            var b = n
            while (true) {
                val r = a + (b - a) / BigInteger.TWO
                val rsq = r * r
                if (rsq <= n && (r + BigInteger.ONE).pow(2) > n) {
                    return r
                } else {
                    if (rsq <= n) {
                        a = r
                    } else {
                        b = r
                    }
                }
            }
        }

        @RequiresApi(Build.VERSION_CODES.TIRAMISU)
        private fun numberToSquareDecomposition(n: BigInteger): List<BigInteger> {
            var num = n - BigInteger.ONE
            val dec = mutableListOf<BigInteger>()
            while (num > BigInteger.ZERO) {
                val lsq = largestSquare(num)
                dec.add(lsq)
                num -= lsq.pow(2)
            }
            return dec
        }

        private fun squareDecompositionToNumber(dec: List<BigInteger>): BigInteger {
            return dec.sumOf { num -> num.pow(2) } + BigInteger.ONE
        }
    }
}

File: webapp/common/bits/bit_tree_converter.js
/*!
 * Copyright (c) 2025 Claycode
 * Licensed under the MIT License with the Commons Clause.
 * Commercial use is prohibited without a separate license.
 * See LICENSE in the project root for details.
 * SPDX-License-Identifier: MIT AND Commons-Clause
 */

import { BitString } from "./bit_string.js";
import { Tree } from "../tree/tree.js";
import { TreeNode } from "../tree/tree_node.js";

const MAX_NUMBER = 2n ** 2000n;

/**
 * Common BitTreeConverter class used by both scanner and generator
 */
export class BitTreeConverter {
    static treeToBits(tree) {
        if (!tree) return new BitString("");

        const number = this.treeToNumber(tree);
        return this.numberToBitString(number);
    }

    static treeToNumber(tree) {
        if (!tree.children || tree.children.length === 0) {
            return 1n;
        }

        const childNumbers = tree.children.map(child => this.treeToNumber(child));

        // Check if any child number is 0 (overflow occurred)
        if (childNumbers.some(num => num === 0n)) {
            console.warn("Overflow occurred in treeToNumber, returning 0");
            return 0n;
        }

        return this.squareDecompositionToNumber(childNumbers);
    }

    static squareDecompositionToNumber(decomposition) {
        let sum = 0n;
        for (const num of decomposition) {
            sum += num * num;

            // Check if sum exceeds MAX_NUMBER
            if (sum + 1n > MAX_NUMBER) {
                return 0n;
            }
        }
        return sum + 1n;
    }

    static numberToBitString(number) {
        const binaryStr = number.toString(2);
        return new BitString(binaryStr.substring(1));
    }

    static bitsToTree(bitsArray) {
        const root = new TreeNode();
        const n = this.bitArrayToInt(bitsArray);
        this.numberToTree(n, root);
        return new Tree(root);
    }

    static bitArrayToInt(bits) {
        // Handle BitString objects
        if (bits instanceof BitString) {
            bits = bits.toString().split('').map(b => parseInt(b));
        }
        // Handle string inputs
        else if (typeof bits === 'string') {
            bits = bits.split('').map(b => parseInt(b));
        }

        let newBits = Array.from(bits);
        newBits.unshift(1); // Add 1 at beginning to make sure it is an integer
        return Array.from(newBits).reverse().reduce((acc, c, i) => acc + BigInt(c) * 2n ** BigInt(i), 0n);
    }

    static intToBitArray(x) {
        if (typeof x !== "bigint") {
            throw new Error("Input must be a BigInt");
        }
        if (x <= 0n) {
            throw new Error("Input must be greater than 0");
        }
        return x.toString(2).slice(1).split("").map((x) => parseInt(x)); // Remove initial 1
    }

    static largestSquareBinsearch(x) {
        if (typeof x !== "bigint") {
            throw new Error("Input must be a BigInt");
        }
        if (x === 1n) {
            return 1n;
        }

        let l = 0n;
        let r = x;

        while (true) {
            const mid = l + ((r - l) / 2n);
            const midSq = mid * mid;

            if (midSq <= x && (mid + 1n) ** 2n > x) {
                return mid;
            } else {
                if (midSq <= x) {
                    l = mid;
                } else {
                    r = mid;
                }
            }
        }
    }

    static numberToSquareDecomposition(x) {
        if (typeof x !== "bigint") {
            throw new Error("Input must be a BigInt");
        }
        const decomposition = [];
        x -= 1n;

        while (x > 0n) {
            const lsq = this.largestSquareBinsearch(x);
            decomposition.push(lsq);
            x -= lsq ** 2n;
        }

        return decomposition;
    }

    // This is needed for the bitsToTree method
    static squareDecompositionToTree(decomposition, root) {
        root.children = decomposition.map(() => new TreeNode());
        for (let i = 0; i < decomposition.length; i++) {
            this.numberToTree(decomposition[i], root.children[i]);
        }
    }

    static numberToTree(n, node) {
        if (n === 1n) return;
        const decomposition = this.numberToSquareDecomposition(n);
        this.squareDecompositionToTree(decomposition, node);
    }
}
